monolith/src/monolith.rs

```diff
@@ -0,0 +1,224 @@
+//! The Monolith-31 permutation.
+//! With significant inspiration from https://extgit.iaik.tugraz.at/krypto/zkfriendlyhashzoo/
+
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::iter;
+use core::marker::PhantomData;
+
+use p3_field::PrimeField32;
+use p3_mds::MdsPermutation;
+use sha3::digest::{ExtendableOutput, Update};
+use sha3::{Shake128, Shake128Reader};
+
+use crate::util::get_random_u32;
+
+// The Monolith-31 permutation.
+// Assumes that F is a 31-bit field (e.g. Mersenne31).
+pub struct Monolith31<F, Mds, const WIDTH: usize, const NUM_ROUNDS: usize>
+where
+    F: PrimeField32,
+    Mds: MdsPermutation<F, WIDTH>,
+{
+    // TODO: if possible, replace with [[F; WIDTH]; NUM_ROUNDS - 1]]
+    pub round_constants: Vec<[F; WIDTH]>,
+    pub lookup1: Vec<u16>,
+    pub lookup2: Vec<u16>,
+    pub mds: Mds,
+
+    _phantom: PhantomData<Mds>,
+}
+
+impl<F, Mds, const WIDTH: usize, const NUM_ROUNDS: usize> Monolith31<F, Mds, WIDTH, NUM_ROUNDS>
+where
+    F: PrimeField32,
```

Comment:
I'm not sure if Monolith can easily generalize to other ~32 bit fields? Particularly the Bars layer seems a bit specific to M31, though I only skimmed it. Maybe we should hardcode `Mersenne31` rather than a generic `F`?

Reply:
Hm yeah I'm fine with that! It seems highly unlikely we'd use a different field even if possible

Reply:
Yeah in general those kinds of lookup-based S-Box are quite field specific. The original Reinforced Concrete gives a highlight for non-friendly fields like BLS12-381 / BN-254, which approach could generalize to other fields by doing an exhaustive search and making sure we don't have too many collisions, but I'd rather hardcode it to M31 here as you suggest.